x-pack/plugins/alerting/server/rules_client/rules_client.ts

```diff
@@ -1258,6 +1350,284 @@ export class RulesClient {
+      throw Error('No rules found for bulk edit');
+    }
+
+    await pMap(
+      buckets,
+      async ({ key: [ruleType, consumer] }) => {
+        this.ruleTypeRegistry.ensureRuleTypeEnabled(ruleType);
+
+        await this.authorization.ensureAuthorized({
+          ruleTypeId: ruleType,
+          consumer,
+          operation: WriteOperations.BulkEdit,
+          entity: AlertingAuthorizationEntity.Rule,
+        });
+      },
+      { concurrency: RULE_TYPE_CHECKS_CONCURRENCY }
+    );
+
+    const { apiKeysToInvalidate, results, errors } = await retryIfBulkEditConflicts(
+      this.logger,
+      `rulesClient.update('operations=${JSON.stringify(options.operations)}, paramsModifier=${
+        options.paramsModifier ? '[Function]' : undefined
+      }')`,
+      (filterKueryNode: KueryNode | null) =>
+        this.bulkEditOcc({
+          filter: filterKueryNode,
+          operations: options.operations,
+          paramsModifier: options.paramsModifier,
+        }),
+      qNodeFilterWithAuth
+    );
+
+    if (apiKeysToInvalidate.length) {
+      await bulkMarkApiKeysForInvalidation(
+        { apiKeys: apiKeysToInvalidate },
+        this.logger,
+        this.unsecuredSavedObjectsClient
+      );
+    }
+
+    const updatedRules = results.map(({ id, attributes, references }) => {
+      return this.getAlertFromRaw<Params>(
+        id,
+        attributes.alertTypeId as string,
+        attributes as RawRule,
+        references,
+        false
+      );
+    });
+
+    return { rules: updatedRules, errors, total };
+  }
+
+  private async bulkEditOcc<Params extends RuleTypeParams>({
+    filter,
+    operations,
+    paramsModifier,
+  }: {
+    filter: KueryNode | null;
+    operations: BulkEditOptions<Params>['operations'];
+    paramsModifier: BulkEditOptions<Params>['paramsModifier'];
+  }): Promise<{
+    apiKeysToInvalidate: string[];
+    rules: Array<SavedObjectsBulkUpdateObject<RawRule>>;
+    resultSavedObjects: Array<SavedObjectsUpdateResponse<RawRule>>;
+    errors: BulkEditError[];
+  }> {
+    const rulesFinder =
+      await this.encryptedSavedObjectsClient.createPointInTimeFinderDecryptedAsInternalUser<RawRule>(
+        {
+          filter,
+          type: 'alert',
+          perPage: 100,
+          ...(this.namespace ? { namespaces: [this.namespace] } : undefined),
+        }
+      );
+
+    const rules: Array<SavedObjectsBulkUpdateObject<RawRule>> = [];
+    const errors: BulkEditError[] = [];
+    const apiKeysToInvalidate: string[] = [];
+    const username = await this.getUserName();
+
+    for await (const response of rulesFinder.find()) {
+      await pMap(
+        response.saved_objects,
+        async (rule) => {
+          try {
+            if (rule.attributes.apiKey) {
+              apiKeysToInvalidate.push(rule.attributes.apiKey);
+            }
+
+            const ruleType = this.ruleTypeRegistry.get(rule.attributes.alertTypeId);
+
+            let attributes = cloneDeep(rule.attributes);
```

Comment:
**question:** is there a place mutating `attributes`? Using `cloneDeep` can be CPU intensive and block the event loop for some time when done at scale.

Reply:
attributes get mutated in `applyBulkEditOperation` when we apply operation to it.

```typescript
            for (const operation of operations) {
              switch (operation.field) {
                case 'actions':
                  await this.validateActions(ruleType, operation.value);
                  ruleActions = applyBulkEditOperation(operation, ruleActions);
                  break;
                default:
                  attributes = applyBulkEditOperation(operation, attributes);
              }
            }
```

At this point the only original part of rule needed here(and it's params):

```typescript
            const validatedMutatedAlertTypeParams = validateMutatedRuleTypeParams(
              validatedAlertTypeParams,
              rule.attributes.params,
              ruleType.validate?.params
            );
```

I think in theory, we can do only copy of  `rule.attributes.params` then. 

What can be implications if we mutate rules.attributes instead of copying them for use in `applyBulkEditOperation` in future? Do you think we should be safe to allow mutating them?